apps/generic-issuance-client/src/pages/pipeline/PipelineEditSection/PipelineActions.tsx

```diff
@@ -0,0 +1,343 @@
+        return;
+      }
+      setActionInProgress(`Deleting pipeline '${pipeline.id}'...`);
+      const res = await deletePipeline(userJWT, pipeline.id);
+      if (res.success) {
+        window.location.href = "/#/dashboard";
+      } else {
+        alert(res.error);
+      }
+    }
+  }, [pipeline.id, setActionInProgress, userJWT]);
+
+  const onRevertToThisVersionClick = useCallback(async () => {
+    if (!historyEntry || !maybeHistoricPipeline) {
+      return;
+    }
+
+    if (
+      !userJWT ||
+      !confirm(
+        `Are you sure you want to revert to version at time ${new Date(
+          historyEntry.timeCreated
+        ).toLocaleString()}? `
+      )
+    ) {
+      return;
+    }
+
+    setActionInProgress(`Reverting pipeline '${pipeline.id}'...`);
+    const historicVersion: Partial<PipelineDefinition> = _.cloneDeep(
+      maybeHistoricPipeline
+    );
+    historicVersion.timeUpdated = pipeline.timeUpdated;
+    const stringifiedDefinition = JSON.stringify(historicVersion);
+    const res = await savePipeline(userJWT, stringifiedDefinition);
+    if (res.success) {
+      window.location.reload();
+    } else {
+      alert(res.error);
+    }
+    setActionInProgress(undefined);
+  }, [
+    historyEntry,
+    maybeHistoricPipeline,
+    pipeline.id,
+    pipeline.timeUpdated,
+    setActionInProgress,
+    userJWT
+  ]);
+
+  const onDuplicateClick = useCallback(async () => {
+    if (
+      !userJWT ||
+      !confirm(
+        "Are you sure you would like to clone this pipeline? " +
+          "It'll be added to your account in a 'paused' state."
+      )
+    ) {
+      return;
+    }
+
+    setActionInProgress(`Duplicating pipeline '${pipeline.id}'...`);
+    const copyDefinition: Partial<PipelineDefinition> = _.cloneDeep(pipeline);
+    delete copyDefinition.id;
+    delete copyDefinition.ownerUserId;
+    copyDefinition.options = {
+      ...copyDefinition.options,
+      paused: true,
+      name: (copyDefinition.options?.name ?? "untitled") + " (copy)"
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } as any;
+
+    const stringifiedDefinition = JSON.stringify(copyDefinition);
+    const res = await savePipeline(userJWT, stringifiedDefinition);
+
+    if (res.success) {
+      window.location.href = "/#/pipelines/" + res.value.id;
+    } else {
+      alert(res.error);
+    }
+    setActionInProgress(undefined);
+  }, [pipeline, setActionInProgress, userJWT]);
+
+  const onUndoClick = useCallback(async () => {
+    if (
+      pipeline &&
+      confirm(
+        "are you sure you want to undo these changes without saving them?"
+      )
+    ) {
+      setEditorValue(stringifyAndFormat(pipeline));
+    }
+  }, [pipeline, setEditorValue]);
+
+  const onSaveClick = useCallback(async () => {
+    if (userJWT && confirm("are you sure you want to update this pipeline?")) {
+      setActionInProgress(`Updating pipeline '${pipeline.id}'...`);
```

Comment:
Does this need to be set back to `undefined` after the action is complete/failed?